docs/source/argument_check.rst

```diff
@@ -0,0 +1,607 @@
+
+Logging
+-------
+
+- There is logging before a quick-return-success or error-return, except when ``handle == nullptr``, when ``rocblas_status_invalid_handle`` is returned, or when ``handle->is_device_memory_size_query()`` returns ``true``.
+
+- Vectors and matrices are logged with their addresses, and are always on device memory.
+
+- Scalar values in device memory are logged as their addresses. Scalar values in host memory are logged as their values, with a ``nullptr`` logged as ``NaN`` (``std::numeric_limits<T>::quiet_NaN()``).
+
+rocBLAS control flow:
+=====================
+
+1. If ``handle == nullptr``, return ``rocblas_status_invalid_handle``.
+
+2. If the function does not require temporary device memory, call the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle);``.
+
+3. If the function requires temporary device memory, and ``handle->is_device_memory_size_query()`` is ``true``, then validate any pointers and arguments required to determine the optimal size of temporary device memory, returning ``rocblas_status_invalid_pointer`` or ``rocblas_status_invalid_size`` if the arguments are invalid, and otherwise ``return handle->set_optimal_device_memory_size(size...);``, where ``size...`` is a list of one or more sizes of temporary buffers, which are allocated with ``handle->device_malloc(size...)`` later.
+
+4. Perform logging if enabled, taking care not to dereference ``nullptr`` arguments.
+
+5. Return rocblas_status_invalid_pointer if any pointers used to determine quick return conditions are NULL.
+
+6. If a scalar argument passed by address is used to determine quick return conditions, and the current handle pointer mode is device, copy the scalar from the device to host, using RETURN_IF_HIP_ERROR() around hipMemcpy(). Otherwise, load the scalar from the host.
+
+5. If quick return conditions are met:
+
+   - If there is a return value
+
+     - For batched or strided_batched, if batch_count < 0, return rocblas_status_invalid_size
+
+     - For batched or strided_batched, if batch_count == 0, return rocblas_status_success
+
+     - If the return value pointer argument is nullptr, return rocblas_status_invalid_pointer
+
+     - Set return value, taking the handle’s pointer mode into account for scalar return values.
+
+     - Return rocblas_status_success.
+
+   - Else,
+
+     -Return rocblas_status_success
+
+6. Check for invalid sizes. Return rocblas_status_invalid_size if size arguments are invalid.
+
+7. Check for NULL pointer arguments not already covered by #3. Return rocblas_status_invalid_pointer if argument pointers are NULL.
+
+8. (Optional.) Allocate device memory, returning rocblas_status_memory_error if the allocation fails.
+
+9. Return rocblas_status_success, assuming no errors in HIP calls or other errors in the calculation.
```

Comment:
Would be good if this section can be reordered to match the general sequence of precedence on checks (if possible?).  i.e. first invalid_size and invalid_value are checked, .. quick return, ... null pointer checks only on data that must be de-referenced based on other values.   Whether the rules can be generalized across all functions we can also summarize.

Reply:
Reordered list